### ISSUE TITLE
(csrf/anti-forgery {:cookie-token true}) returns 404 if session isn't set

### ISSUE THREAD

**Issue by grzm** [pedestal/pedestal#459]
# Description

When using `(csrf/anti-forgery {:cookie-token true})`, a request that does not include a token gets a 404 Not Found response rather than a 403 Forbidden.

## Expected Behavior

If the request does not include a valid CSRF token, the response should be 403 Forbidden.

## Actual Behavior

A request without a valid CSRF token receives a 404 Not Found response.

## Steps to reproduce

Test case repo: https://github.com/grzm/pedestal-csrf-anti-forgery-cookie-404

In brief:
```clojure
(defn home-page
  [request]
  (ring-resp/response "Hello World!\n"))

(def common-interceptors [middlewares/cookies
                          (middlewares/session {})
                          (body-params/body-params)
                          http/html-body])

(def csrf-interceptors [middlewares/cookies
                        (middlewares/session {})
                        (csrf/anti-forgery)
                        (body-params/body-params)
                        http/html-body])

(def csrf-cookie-interceptors [middlewares/cookies
                               (middlewares/session {})
                               (csrf/anti-forgery {:cookie-token true})
                               (body-params/body-params)
                               http/html-body])

(def routes #{["/csrf" :post (conj csrf-interceptors `home-page)
               :route-name :post-csrf]
              ["/csrf-cookie" :post (conj csrf-cookie-interceptors `home-page)
               :route-name :post-csrf-cookie]})
```

```
λ curl -v localhost:8080/csrf -X POST
*   Trying ::1...
* Connected to localhost (::1) port 8080 (#0)
> POST /csrf HTTP/1.1
> Host: localhost:8080
> User-Agent: curl/7.46.0
> Accept: */*
> 
< HTTP/1.1 403 Forbidden
< Date: Mon, 29 Aug 2016 15:09:58 GMT
< Strict-Transport-Security: max-age=31536000; includeSubdomains
< X-Frame-Options: DENY
< X-Content-Type-Options: nosniff
< X-XSS-Protection: 1; mode=block
< Content-Type: text/html
< Transfer-Encoding: chunked
< Server: Jetty(9.3.8.v20160314)
< 
* Connection #0 to host localhost left intact
<h1>Invalid anti-forgery token</h1>


λ curl -v localhost:8080/csrf-cookie -X POST
*   Trying ::1...
* Connected to localhost (::1) port 8080 (#0)
> POST /csrf-cookie HTTP/1.1
> Host: localhost:8080
> User-Agent: curl/7.46.0
> Accept: */*
> 
< HTTP/1.1 404 Not Found
< Date: Mon, 29 Aug 2016 15:09:53 GMT
< Content-Type: text/plain
< Transfer-Encoding: chunked
< Server: Jetty(9.3.8.v20160314)
< 
* Connection #0 to host localhost left intact
Not Found
```
## Operating System (including version).
```
λ sw_vers
ProductName:	Mac OS X
ProductVersion:	10.10.5
BuildVersion:	14F1909
```

## Your current Leiningen version (`lein --version`)

```
λ lein -v
Leiningen 2.6.1 on Java 1.8.0_45 Java HotSpot(TM) 64-Bit Server VM
```
## Pedestal version

0.5.1 and 0.5.2-SNAPSHOT

**Comment by grzm** [comment]
I believe this is because `assoc-double-submit-cookie` returns `nil` if there is no session set. Should it just inspect the response and if it's 403 just pass that through?

**Comment by ohpauleez** [comment]
@grzm - Are you using *any* session at all (like a session cookie)?  CSRF can only work is there is a session in place.  It's not secure enough if you just relied on a cookie outside of session alone.  The cookie is *only* there to allow rich client-side apps to make CSRF requests *for a given session*, otherwise an attacker could use a stale cookie to launch attacks.

The CSRF comments also state the session needs to be processed first, and if loaded through the default interceptors, CSRF automatically [enables sessions](https://github.com/pedestal/pedestal/blob/master/service/src/io/pedestal/http.clj#L247)

All of that aside, putting an `if` there instead of a when, to forward would fix the 404 (which is caused because the response turns into `nil`

**Comment by grzm** [comment]
In the test case I provided, session is included in the interceptor chain but the end-handler isn't doing anything with it. I know this isn't enough on its own to ensure security. However, the behavior of returning a 404 instead of a 403 even if things aren't completely set up correctly is quite surprising if not wrong. The pull request I submitted does replace the `when` with an `if` and returns the 403 as expected.
